tgbot/src/webhook.rs

```diff
@@ -0,0 +1,165 @@
+use crate::handler::UpdateHandler;
+use bytes::buf::BufExt;
+use futures_util::future::{ok, ready, FutureExt, Ready};
+use http::Error as HttpError;
+use hyper::{body, service::Service, Body, Error as HyperError, Method, Request, Response, Server, StatusCode};
+use log::error;
+use std::{
+    convert::Infallible,
+    error::Error as StdError,
+    future::Future,
+    net::SocketAddr,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::sync::Mutex;
+
+#[doc(hidden)]
+pub struct WebhookServiceFactory<H> {
+    path: String,
+    handler: Arc<Mutex<H>>,
+}
+
+impl<H> WebhookServiceFactory<H> {
+    #[doc(hidden)]
+    pub fn new<P>(path: P, update_handler: H) -> Self
+    where
+        P: Into<String>,
+    {
+        WebhookServiceFactory {
+            path: path.into(),
+            handler: Arc::new(Mutex::new(update_handler)),
+        }
+    }
+}
+
+impl<H, T> Service<T> for WebhookServiceFactory<H> {
+    type Response = WebhookService<H>;
+    type Error = Infallible;
+    type Future = Ready<Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _: &mut Context) -> Poll<Result<(), Self::Error>> {
+        Ok(()).into()
+    }
+
+    fn call(&mut self, _: T) -> Self::Future {
+        let path = self.path.clone();
+        ok(WebhookService {
+            path,
+            handler: self.handler.clone(),
+        })
+    }
+}
+
+#[doc(hidden)]
+pub struct WebhookService<H> {
+    path: String,
+    handler: Arc<Mutex<H>>,
```

Comment:
Get rid of `Mutex`

Reply:
Handler needs to take &mut self, so we can't